### ISSUE TITLE
Issue comparing Expressions when one value is a constant

### ISSUE THREAD

**Issue by QTom01** [moq/moq4#1054]
I have discovered an issue when using Expressions in a Setup on a mock object.

It seems using a simple comparison Expression fails to match properly when comparing to a string constant.

The below code should reproduce it:
```
[Test]
public async Task ExpressionTest()
{
    var s1 = "example";

    // Does not work
    const string s2 = "example";

    // Works
    //var s2 = "example";

    var mock = new Mock<IMockObject>();

    mock.Setup(x => x.Method(y => y.Id == s1))
        .ReturnsAsync("mockResult");

    var result = await mock.Object.Method(x => x.Id == s2);
    // Exprect result to be "mockResult", instead it is null

    Assert.AreEqual(result, "mockResult");
}

public class MyObject
{
    public string Id { get; set; }
}

public interface IMockObject
{
    public Task<string> Method(Expression<Func<MyObject, bool>> expression);
}

```

The two strings s1 and s2 are the same, however when one of them is a constant then it fails to match the two expressions.

I would expect it should work the same when the string is a constant?

This is on a .NET Core 3.1 test project using Moq 4.14.5 and NUnit 3.12.

**Comment by QTom01** [comment]
For now I can work around by simply assigning a local variable in the test from the constant and using that instead.

**Comment by stakx** [comment]
Referencing a constant vs. a variable results in slightly different LINQ expression trees: in the former case, the string value gets embedded as a constant, while in the latter case, the variable gets captured in the form of a reference to an anonymous object's field.

We've had a few similar issues in the past, and the approach followed most recently was that captured variables should be evaluated down to a constant value at the time when two LINQ expression trees get compared.

If we applied this rule here, this present issue would be resolved.

However, at present I am not sure whether this rule *should* apply in this case. The added complication here is that we have a setup LINQ expression tree that embeds another LINQ expression tree (the argument passed to `Method`). There are two principal ways how expression trees reference other expression trees: [`Expression.Constant` and `Expression.Quote`](https://stackoverflow.com/q/3716492). I'm not positively certain that Moq makes this distinction correctly in all situations.

TL;DR: I'll look into it, but this will take some thought.
